Ban type alias names ending in "T"?

I don't like type aliases in a stub to have names ending with "T". For me, it's confusing -- for me, the "T" suffix implies that the object is a `TypeVar` rather than a type alias.

What do we think about a flake8-pyi check to ban type aliases from having names ending in "T"? I've got [a proof of concept on my fork](https://github.com/PyCQA/flake8-pyi/compare/master...AlexWaygood:alias-names), but I figured I'd see what other people think before writing tests and filing a PR, since it's quite opinionated.

The hits from my proof-of-concept are:

```
./stubs/redis/redis/typing.pyi:17:1: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/tracemalloc.pyi:45:1: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/tracemalloc.pyi:65:5: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/tracemalloc.pyi:67:5: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/_ast.pyi:369:5: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/_ast.pyi:372:5: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/_codecs.pyi:111: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
./stdlib/asyncio/tasks.pyi:61:1: Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
```

Of these, the `redis` hit is a false-positive: the type alias exists at runtime, and the name at runtime ends with "T". But all the others are true positives.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -11,6 +11,7 @@ Features:
   emits an error for assignments to `typing.Annotated`, `typing.Optional` and
   `typing.Any`, as well as subscripted `tuple`s, `dict`s, `set`s, `frozenset`s,
   `list`s, and `type`s.
+* Introduce Y043: Ban type aliases from having names ending with an uppercase "T".
 * Support Python 3.11.
 * Support `typing_extensions.overload` and `typing_extensions.NamedTuple`. (The latter
   is not yet released, but will be in the next version of `typing_extensions`.)

--- a/README.md
+++ b/README.md
@@ -73,6 +73,7 @@ currently emitted:
 | Y039 | Use `str` instead of `typing.Text`. This error code is incompatible with stubs supporting Python 2.
 | Y040 | Never explicitly inherit from `object`, as all classes implicitly inherit from `object` in Python 3. This error code is incompatible with stubs supporting Python 2.
 | Y041 | Y041 detects redundant numeric unions. For example, PEP 484 specifies that type checkers should treat `int` as an implicit subtype of `float`, so `int` is redundant in the union `int \| float`. In the same way, `int` is redundant in the union `int \| complex`, and `float` is redundant in the union `float \| complex`.
+| Y043 | Do not use names ending in "T" for private type aliases. (The "T" suffix implies that an object is a `TypeVar`.)
 
 Many error codes enforce modern conventions, and some cannot yet be used in
 all cases:

--- a/pyi.py
+++ b/pyi.py
@@ -917,6 +917,13 @@ class PyiVisitor(ast.NodeVisitor):
         else:
             self.generic_visit(node)
 
+    # Y043: Error for alias names in "T"
+    # (plus possibly a single digit afterwards), but only if:
+    #
+    # - The name starts with "_"
+    # - The penultimate character in the name is an ASCII-lowercase letter
+    _Y043_REGEX = re.compile(r"^_.*[a-z]T\d?$")
+
     def visit_AnnAssign(self, node: ast.AnnAssign) -> None:
         node_annotation = node.annotation
         if _is_Final(node_annotation):
@@ -939,6 +946,10 @@ class PyiVisitor(ast.NodeVisitor):
         if _is_TypeAlias(node_annotation):
             with self.visiting_TypeAlias.enabled():
                 self.generic_visit(node)
+            if isinstance(node_target, ast.Name) and self._Y043_REGEX.match(
+                target_name
+            ):
+                self.error(node, Y043)
             return
 
         self.generic_visit(node)
@@ -1678,3 +1689,4 @@ Y038 = 'Y038 Use "from collections.abc import Set as AbstractSet" instead of "fr
 Y039 = 'Y039 Use "str" instead of "typing.Text"'
 Y040 = 'Y040 Do not inherit from "object" explicitly, as it is redundant in Python 3'
 Y041 = 'Y041 Use "{implicit_supertype}" instead of "{implicit_subtype} | {implicit_supertype}" (see "The numeric tower" in PEP 484)'
+Y043 = 'Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)'

--- a/tests/aliases.pyi
+++ b/tests/aliases.pyi
@@ -58,3 +58,12 @@ class Foo:
 
 f: Foo = ...
 baz = f.baz
+
+_PrivateAliasT: TypeAlias = str | int  # Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
+_PrivateAliasT2: TypeAlias = typing.Any  # Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
+_PrivateAliasT3: TypeAlias = Literal["not", "a", "chance"]  # Y043 Bad name for a type alias (the "T" suffix implies a TypeVar)
+PublicAliasT: TypeAlias = str | int
+PublicAliasT2: TypeAlias = Union[str, bytes]
+_ABCDEFGHIJKLMNOPQRST: TypeAlias = typing.Any
+_PrivateAliasS: TypeAlias = Literal["I", "guess", "this", "is", "okay"]
+_PrivateAliasS2: TypeAlias = Annotated[str, "also okay"]

